apps/snitch_core/test/snitch/data/model/stock/stock_location_test.exs

```diff
@@ -0,0 +1,145 @@
+defmodule Snitch.Data.Model.StockLocationTest do
+  use ExUnit.Case, async: true
+  use Snitch.DataCase
+  import Snitch.Factory
+  alias Snitch.Data.Model
+
+  describe "create/4" do
+    test "Fails for blank attributes" do
+      assert {:error, changeset} = Model.StockLocation.create("", "", nil, nil)
+
+      assert [
+               name: {"can't be blank", [validation: :required]},
+               address_line_1: {"can't be blank", [validation: :required]},
+               state_id: {"can't be blank", [validation: :required]},
+               country_id: {"can't be blank", [validation: :required]}
+             ] = changeset
+    end
+
+    test "Fails for invalid associations" do
+      assert {:error, changeset} = Model.StockLocation.create("London", "Digon Alley", 1, 1)
+      assert %{state_id: ["does not exist"]} = errors_on(changeset)
+
+      state = insert(:state)
+
+      assert {:error, changeset} =
+               Model.StockLocation.create("London", "Digon Alley", state.id, 1)
+
+      assert %{country_id: ["does not exist"]} = errors_on(changeset)
+    end
+
+    test "Inserts with valid attributes" do
+      assert {:ok, stock_location} =
+               Model.StockLocation.create(
+                 "London",
+                 "Digon Alley",
+                 insert(:state).id,
+                 insert(:country).id
+               )
+    end
+  end
+
+  describe "get/1" do
+    test "Fails with invalid id" do
+      stock_location = Model.StockLocation.get(1)
+      assert nil == stock_location
+    end
+
+    test "gets with valid id" do
+      insert_stock_location = insert(:stock_location)
+
+      get_stock_location = Model.StockLocation.get(insert_stock_location.id)
+      assert insert_stock_location.id == get_stock_location.id
+      assert insert_stock_location.name == get_stock_location.name
+
+      # with stock location map
+      get_stock_location = Model.StockLocation.get(%{id: insert_stock_location.id})
+      assert insert_stock_location.id == get_stock_location.id
+      assert insert_stock_location.id == get_stock_location.id
+      assert insert_stock_location.name == get_stock_location.name
+    end
+  end
+
+  describe "update/1" do
+    test "Fails for INVALID attributes" do
+      stock_location = insert(:stock_location)
+
+      assert {:error, changeset} =
+               Model.StockLocation.update(%{name: "", address_line_1: "", id: stock_location.id})
+
+      assert [
+               name: {"can't be blank", [validation: :required]},
+               address_line_1: {"can't be blank", [validation: :required]}
+             ] = changeset
+    end
+
+    test "updates for VALID attributes" do
+      stock_location = insert(:stock_location)
+
+      assert {:ok, updated_stock_location} =
+               Model.StockLocation.update(%{name: "Updated New", id: stock_location.id})
+
+      assert stock_location.name != updated_stock_location.name
+    end
+  end
+
+  describe "update/2" do
```

Comment:
Actually I thought of it as update with 2 params where next one can be nil. so wanted to cover that. This can be done under one describe block though